galley/pkg/config/analysis/analyzers/deprecation/deprecation.go

```diff
@@ -0,0 +1,114 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package deprecation
+
+import (
+	"istio.io/api/networking/v1alpha3"
+
+	"istio.io/istio/galley/pkg/config/analysis"
+	"istio.io/istio/galley/pkg/config/analysis/msg"
+	"istio.io/istio/galley/pkg/config/collection"
+	"istio.io/istio/galley/pkg/config/processor/metadata"
+	"istio.io/istio/galley/pkg/config/resource"
+)
+
+// FieldAnalyzer checks for deprecated Istio types and fields
+type FieldAnalyzer struct{}
```

Comment:
@sushicw 
It may make sense to aggregate this with the VirtualService analyzer (either within the same analyzer, or as a sibling of it under the same package). 
Alternatively, you can make this specific to VirtualService only. (i.e. deprecated.VirtualServiceAnalyzer).

Reason being: the analyzers should be scoped/packaged such that they can be enabled/disabled individually, depending on the availability of resources or user selection. Having an analyzer that will cross cut and touch potentially every resource means, the analyzer may get disabled if one of the resources mentioned is not available.

Reply:
I agree. 

In the longer run I expect this actually may make more sense packaged alongside the single-object validation (most of which currently lives under istio/pkg/config/validation/validation.go). This PR does demonstrate that the new framework makes it easy to write single-object validation rules! 

I'm also very curious to know if deprecated fields can be validated using Cue: @jasonwzm any comment there? I know that Jason had been planning to replace much (if not all) of the current validation.go code with rules generated via Cue.

Reply:
@sushicw `istioctl validate` calls the same `validation.ValidateFunc` as the webhook.

Consider setting up analyzers to call the existing ValidateFuncs.  Under normal circumstances these should not trigger errors because the webhook does not let invalid objects in.  Do this for analyzing before applying to the cluster, or in case the web hook was disabled.

The reason I wrote the deprecation in the analysis framework is that validation.ValidateFunc doesn't offer a way to surface info and warning level messages.  The webhook must allow deprecated fields in.

Reply:
To follow up on Oz's original point here: I think we're OK for now, since I don't think we currently have any toggles that would disable Istio CRDs for validation. If/when that happens, we can split this analyzer up. (Unless by then we've already switched to a different model, in which case it wouldn't matter).